pyccel/ast/datatypes.py

```diff
@@ -197,6 +199,55 @@ def shape_is_compatible(self, shape):
         """
         return shape is None
 
+#==============================================================================
+class FinalType:
+    """
+    A class to get PyccelType subclasses describing constant values.
+
+    A class to get PyccelType subclasses describing constant values.
+    """
+    __slots__ = ()
+
+    @classmethod
+    @lru_cache
+    def get_new(cls, underlying_type):
+        """
+        Get the parameterised Final type.
+
+        Get the parameterised Final type Final[underlying_type].
+
+        Parameters
+        ----------
+        underlying_type : PyccelType
+            The type which is characterised as final.
+        """
+        if isinstance(underlying_type, FinalType):
+            return underlying_type
+
+        type_class = type(underlying_type)
+        def __init__(self):
+            self._underlying_type = underlying_type
+            type(underlying_type).__init__(self)
+        def __hash__(self):
+            return type_class.__hash__(underlying_type)
+        def __eq__(self, other):
+            return type_class.__eq__(underlying_type, other)
+        def get_underlying_type(self):
+            """
+            Get the type that is indicated as const.
+
+            Get the type that is indicated as const.
+            """
+            return self._underlying_type
+        return type(f'Final{type_class.__name__}', (FinalType, type_class,),
+                    {'__init__' : __init__,
+                     '__hash__' : __hash__,
+                     '__eq__' : __eq__,
+                     'underlying_type': property(get_underlying_type)})()
```

Comment:
Are we not missing the square brackets in the type name?

```suggestion
        return type(f'Final[{type_class.__name__}]', (FinalType, type_class,),
                    {'__init__' : __init__,
                     '__hash__' : __hash__,
                     '__eq__' : __eq__,
                     'underlying_type': property(get_underlying_type)})()
```

Further, when I try to use this constructor passing it a subclass of `PyccelType` I get this error:
```python
TypeError: type.__new__() takes exactly 3 arguments (0 given)
```

Reply:
Done [306609b](https://github.com/pyccel/pyccel/pull/2415/commits/306609b415f7ed71b116144f01d7410e588445e4)

Reply:
Me neither! Apparently any string is accepted...